core/modules/wsched/ChunkTaskList.cc

```diff
@@ -0,0 +1,393 @@
+                    str += tblInfo.tableName + " ";
+                }
+                LOGS(_log, LOG_LVL_DEBUG, "ready memMan " << msg << " - " << str);
+            }
+        };
+    if (_readyTask != nullptr) {
+        return ChunkTasks::ReadyState::READY;
+    }
+    if (_activeTasks.empty()) {
+        return ChunkTasks::ReadyState::NOT_READY;
+    }
+    // Calling this function doesn't get expensive until it gets here. Luckily,
+    // after this point it will return READY or NO_RESOURCES, and ChunkTaskList::_ready
+    // will not examine any further chunks upon seeing those results.
+    auto task = _activeTasks.top();
+    if (!task->hasMemHandle()) {
+        memman::TableInfo::LockType lckOptTbl = memman::TableInfo::LockType::MUSTLOCK;
+        memman::TableInfo::LockType lckOptIdx = memman::TableInfo::LockType::NOLOCK;
+        if (useFlexibleLock) lckOptTbl = memman::TableInfo::LockType::FLEXIBLE;
+        auto scanInfo = task-> getScanInfo();
+        auto chunkId = task->getChunkId();
+        if (chunkId != _chunkId) {
+            // This would slow things down badly, but the system would survive.
+            LOGS(_log, LOG_LVL_ERROR, "ChunkTasks " << _chunkId << " got task for chunk " << chunkId
+                    << " " << task->getIdStr());
+        }
+        std::vector<memman::TableInfo> tblVect;
+        for (auto const& tbl : scanInfo.infoTables) {
+            memman::TableInfo ti(tbl.db + "/" + tbl.table, lckOptTbl, lckOptIdx);
+            tblVect.push_back(ti);
+        }
+        // If tblVect is empty, we should get the empty handle
+        memman::MemMan::Handle handle = _memMan->lock(tblVect, chunkId);
+        if (handle == 0) {
+            switch (errno) {
+            case ENOMEM:
+                logMemManRes(true, "ENOMEM", tblVect);
+                return ChunkTasks::ReadyState::NO_RESOURCES;
+            case ENOENT:
+                LOGS(_log, LOG_LVL_ERROR, "_memMgr->lock errno=ENOENT chunk not found " << task->getIdStr());
+                // Not sure if this is the best course of action, but it should just need one
+                // logic path. The query should fail from the missing tables
+                // and the czar must be able to handle that with appropriate retries.
+                handle = memman::MemMan::HandleType::ISEMPTY;
+                break;
+            default:
+                LOGS(_log, LOG_LVL_ERROR, "_memMgr->lock file system error " << task->getIdStr());
+                // Any error reading the file system is probably fatal for the worker.
+                throw std::bad_exception();
+                return ChunkTasks::ReadyState::NO_RESOURCES;
+            }
+        }
+        task->setMemHandle(handle);
+        logMemManRes(false, task->getIdStr() + " got handle", tblVect);
+    }
+
+    // There is a Task to run at this point, pull it off the heap to avoid confusion.
+    auto popped = _activeTasks.pop();
+    _readyTask = task;
+    if (popped != task) {
```

Comment:
Hmm, how could this ever happen? I guess I'm always suspicious when I see a throw. I would recommend an alternate recovery action other than killing the whole server (I didn't see a catch, was there one?). Perhaps just abandon the task and keep going.